state/trackableDataTrie.go

```diff
@@ -144,129 +238,122 @@ func (tdaw *trackableDataTrie) SaveDirtyData(mainTrie common.Trie) ([]common.Tri
+			Value:   dataEntry.value,
+			Version: dataEntry.oldVersion.version,
 		}
+	}
 
-		oldValues[index] = oldEntry
-
-		err = tdaw.updateValInTrieWithAutoBalancing([]byte(key), val, dtr)
-		if err != nil {
-			return nil, err
+	if tdaw.enableEpochsHandler.IsAutoBalanceDataTriesEnabled() {
+		hashedKey := tdaw.hasher.Compute(string(key))
+		oldVal, _, err := tdaw.tr.Get(hashedKey)
+		if err == nil && len(oldVal) != 0 {
+			return core.TrieData{
+				Key:     hashedKey,
+				Value:   oldVal,
+				Version: core.AutoBalanceEnabled,
+			}
 		}
+	}
 
-		index++
+	oldVal, _, err := tdaw.tr.Get(key)
+	if err == nil && len(oldVal) != 0 {
+		return core.TrieData{
+			Key:     key,
+			Value:   oldVal,
+			Version: core.NotSpecified,
+		}
 	}
 
-	tdaw.dirtyData = make(map[string][]byte)
-	return oldValues, nil
+	newDataVersion := tdaw.getVersionForNewlyAddedData()
+	return core.TrieData{
+		Key:     tdaw.getKeyForVersion(key, newDataVersion),
+		Value:   nil,
+		Version: newDataVersion,
+	}
 }
 
-func (tdaw *trackableDataTrie) getOldKeyAndValWithCleanup(key string) (common.TrieData, error) {
-	hashedKey := tdaw.hasher.Compute(key)
+func (tdaw *trackableDataTrie) deleteOldEntryIfMigrated(key []byte, newData dirtyData, oldEntry core.TrieData) error {
+	if !tdaw.enableEpochsHandler.IsAutoBalanceDataTriesEnabled() {
+		return nil
+	}
 
-	oldVal, _, err := tdaw.tr.Get(hashedKey)
-	if err == nil && len(oldVal) != 0 {
-		return common.TrieData{
-			Key:     hashedKey,
-			Value:   oldVal,
-			Version: common.AutoBalanceEnabled,
-		}, nil
+	if oldEntry.Version == core.NotSpecified && newData.newVersion == core.AutoBalanceEnabled {
+		return tdaw.tr.Delete(key)
 	}
 
-	oldVal, _, err = tdaw.tr.Get([]byte(key))
+	return nil
+}
+
+func (tdaw *trackableDataTrie) modifyTrie(key []byte, dataEntry dirtyData, oldVal core.TrieData, dtr dataTrie) error {
+	shouldInsertInTrie, err := tdaw.deleteFromTrieIfNilVal(dataEntry.value, oldVal, key, dtr)
 	if err != nil {
-		return common.TrieData{}, err
+		return err
 	}
-
-	if len(oldVal) == 0 {
-		return common.TrieData{
-			Key:     hashedKey,
-			Value:   nil,
-			Version: common.NotSpecified,
-		}, nil
+	if !shouldInsertInTrie {
+		return nil
 	}
 
-	err = tdaw.tr.Delete([]byte(key))
+	newKey := tdaw.getKeyForVersion(key, dataEntry.newVersion)
+	value, err := tdaw.getValueForVersion(key, dataEntry.value, dataEntry.newVersion)
 	if err != nil {
-		return common.TrieData{}, err
+		return err
 	}
+	version := dataEntry.newVersion
 
-	return common.TrieData{
-		Key:     []byte(key),
-		Value:   oldVal,
-		Version: common.NotSpecified,
-	}, nil
+	return dtr.UpdateWithVersion(newKey, value, version)
 }
 
-func (tdaw *trackableDataTrie) updateValInTrieWithAutoBalancing(key []byte, val []byte, selfDataTrie dataTrie) error {
-	if len(val) == 0 {
-		return tdaw.tr.Delete(tdaw.hasher.Compute(string(key)))
+func (tdaw *trackableDataTrie) deleteFromTrieIfNilVal(val []byte, oldVal core.TrieData, key []byte, dtr dataTrie) (bool, error) {
```

Comment:
not clear what the return values are for
why do we return true on len(val) != 0 and return false on len(oldValue.Value) == 0 ?
what is the logic for that?

maybe add a comment & give names to return values?